### PR TITLE
MatrixOfConstraints

### PR DESCRIPTION
**Pull request by blegat**
Current benchmark:
```julia
import MathOptInterface
const MOI = MathOptInterface
const MOIU = MOI.Utilities
import Clp
function bench(n, m)
    model = MOIU.Model{Float64}()
    x = MOI.add_variables(model, n)
    fx = MOI.SingleVariable.(x)
    ofx = 1.0 .* fx
    for i in 1:n
        MOI.add_constraint(model, ofx[i], MOI.LessThan(1.0))
        MOI.add_constraint(model, ofx[i], MOI.GreaterThan(1.0))
    end
    s = sum(ofx)
    for i in 1:n
        MOI.add_constraint(model, s, MOI.EqualTo(1.0))
    end
    optimizer = MOI.Bridges.full_bridge_optimizer(
        MOIU.CachingOptimizer(
            M(),
            Clp.Optimizer(),
        ),
        Float64,
    )
    @time MOI.copy_to(optimizer, model)
    @time MOIU.attach_optimizer(optimizer.model)
    optimizer.model
end
```

With this branch, Cbc#od/moi10 and `M=MOIU.Model{Float64}`, I get:
```
julia> bench(1000, 1000);
  0.035744 seconds (18.71 k allocations: 32.124 MiB)
  0.075348 seconds (429 allocations: 42.859 MiB)
```

With this branch, Cbc#bl/matrix and `M=Cbc.Model`, I get:
```
julia> bench(1000, 1000);
  0.011251 seconds (698 allocations: 551.617 KiB)
  0.000164 seconds (105 allocations: 146.023 KiB)
```

What really matters is the sum of the two. The speedup is approximately 10x.
~~We already have a significant decrease but I think we could have a number of allocation which does not scale linearly with `n` and `m` so I'm trying to figure it out.~~
The number of allocations is a constant + some logarithmic on `n` and `m` which is expected (e.g. because of `push` in `caches`).

See #1261

Requires #1267, #1286

### CODE REVIEW

**Comment by odow** [comment]
What's the status of this? Can we make a new MOI release without it?

**Comment by blegat** [comment]
There is still some work needed, we don't need to wait for it for a new MOI release

File: src/Test/config.jl [hunk 1]
```diff
@@ -2,6 +2,7 @@ struct TestConfig{T<:Real}
     atol::T # absolute tolerance for ...
     rtol::T # relative tolerance for ...
     solve::Bool # optimize and test result
+    query_number_of_constraints::Bool # can get `MOI.NumberOfConstraints` attribute
```
Comment (odow) [inline src/Test/config.jl:5]: This should be a separate PR.

**Comment by odow** [comment]
As another benchmark: how is the first solve? Compilation times are our big killer for this.

File: src/Test/config.jl (hunk 1 above)
Comment (blegat) [reply src/Test/config.jl:5]: Done in https://github.com/jump-dev/MathOptInterface.jl/pull/1327

File: src/Utilities/copy.jl [hunk 2]
```diff
@@ -69,6 +69,25 @@ function IndexMap(n = 0)
         DoubleDicts.IndexDoubleDict(),
     )
 end
+function _identity_constraints_map(
+    index_map::MOIU.DoubleDicts.IndexWithType{F,S},
```
Comment (odow) [inline src/Utilities/copy.jl:73]: Same comment about argument order. Model first?

File: src/Utilities/copy.jl [hunk 3]
```diff
@@ -608,6 +627,8 @@ function try_constrain_variables_on_creation(
     single_variable_not_added
 end
 
+function final_touch(::MOI.ModelLike, idxmap) end
```
Comment (odow) [inline src/Utilities/copy.jl:630]: This needs a docstring.

File: src/Utilities/matrix_of_constraints.jl [hunk 4]
```diff
@@ -0,0 +1,614 @@
+abstract type ProductOfSets{T} end
```
Comment (odow) [inline src/Utilities/matrix_of_constraints.jl:1]: This needs a docstring explaining how it stores the constraints.

File: src/Utilities/model.jl [hunk 5]
```diff
@@ -14,14 +14,21 @@ function MOI.get(model::AbstractModel, ::MOI.NumberOfVariables)::Int64
...
     vi = VI(model.num_variables_created += 1)
     push!(model.single_variable_mask, 0x0)
-    push!(model.lower_bound, zero(T))
-    push!(model.upper_bound, zero(T))
+    push!(model.lower_bound, _no_lower_bound(T))
+    push!(model.upper_bound, _no_upper_bound(T))
     if model.variable_indices !== nothing
         push!(model.variable_indices, vi)
     end
+    _add_variable(model.constraints)
```
Comment (odow) [inline src/Utilities/model.jl:31]: It seems weird to force other types to overload `_add_variable`. Why not just overload `MOI.add_variable(model.constraints)`?

**Review by odow — COMMENTED** [review]
This is quite hard to review because there is a lot of new types being added
- ProductOfSets
- MixOfScalarSets
- OrderedProductOfSets
- OrderedProductOfScalarSets
- UnevenIterator
- Box
- MatrixOfConstraints
- IdentityMap
- MutableSparseMatrixCSC

Can we break it down into smaller pieces? It's not obvious why we need so many different structures, or how they inter-relate.

File: src/Utilities/model.jl (hunk 5 above)
Comment (blegat) [reply src/Utilities/model.jl:31]: Then `VectorOfConstraints` should return a `VariableIndex` but currently it does not keep track of this information.

File: src/Utilities/copy.jl (hunk 3 above)
Comment (blegat) [reply src/Utilities/copy.jl:572]: Let me know what you think for the name. Another option would be `end_of_copy`

File: test/Utilities/copy.jl [hunk 6]
```diff
@@ -653,6 +653,7 @@ struct OnlyCopyConstraints{F,S} <: MOI.ModelLike
         return new{F,S}(MOIU.VectorOfConstraints{F,S}())
     end
 end
+function MOI.Utilities._add_variable(::OnlyCopyConstraints) end
```
Comment (odow) [inline test/Utilities/copy.jl:656]: If this is a public API that people need to implement, it shouldn't begin with `_`.

File: src/Utilities/matrix_of_constraints.jl [hunk 7]
```diff
@@ -0,0 +1,357 @@
+# Constants: stored in a `Vector` or `Box` or any other type implementing:
+# `empty!`, `resize!` and `load_constants`.
+
+"""
+    load_constants(constants, offset, func_or_set)
+
+This function loads the constants of `func_or_set` in `constants` at an offset
+of `offset`. The storage should be preallocated with `resize!` before calling
+this function.
```
Comment (odow) [inline src/Utilities/matrix_of_constraints.jl:9]: What is `offset`? "Where `offset` is the number of constants that have previously been loaded"?

File: src/Utilities/matrix_of_constraints.jl [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+    empty!(b.lower)
+    empty!(b.upper)
+    return b
+end
+function Base.resize!(b::Box, n)
+    resize!(b.lower, n)
+    resize!(b.upper, n)
+    return
+end
+function load_constants(b::Box{T}, offset, set) where {T}
```
Comment (odow) [inline src/Utilities/matrix_of_constraints.jl:47]: Should `set` have some type restrictions here?

File: src/Utilities/matrix_of_constraints.jl [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+
+"""
+    rows(model::MatrixOfConstraints, ci::MOI.ConstraintIndex)
+
+Return the rows corresponding to the constraint of index `ci`. If it is a
+vector constraint, this is a `UnitRange`, otherwise, this is an integer.
+"""
+rows(model::MatrixOfConstraints, ci::MOI.ConstraintIndex) = rows(model.sets, ci)
+
+function affine_function_type(
```
Comment (odow) [inline src/Utilities/matrix_of_constraints.jl:137]: Is this a public function? If not, it should start with `_`.

File: src/Utilities/matrix_of_constraints.jl [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+    ::Type{T},
+    ::Type{<:MOI.AbstractScalarSet},
+) where {T}
+    return MOI.ScalarAffineFunction{T}
+end
+function affine_function_type(
```
Comment (odow) [inline src/Utilities/matrix_of_constraints.jl:143]: A pet peeve of mine: new lines between methods?

File: src/Utilities/matrix_of_constraints.jl [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+a `src::MOI.Utilities.Model` to a bridged model with a `MatrixOfConstraints`.
+For all the types that are not bridged, the constraints will be copied
+with `pass_nonvariable_constraints` hence the functions stored in
+`caches` are exactly the same as the ones stored in `src`.
+This is ok since this is only during the `copy_to` operation during which `src`
+cannot be modified.
+On the other hand, for the types that are bridged, the functions added
+may contain duplicates even if the functions did not contain duplicates in
+`src` so duplicates are removed with `MOI.Utilities.canonical`.
+"""
```
Comment (odow) [inline src/Utilities/matrix_of_constraints.jl:106]: I started making inline notes, but I'll stop. The API needed to implement MatrixOfConstraints is quite large, and needs to be documented.

This docstring, plus a page in the documentation, should mention what functions need to be implemented for the different `coefficients`, `constants`, and `sets` (e.g., `set_types`, `rows`, ...) fields.


See the comment at the top of this file, and the notes in `ProductOfSets.jl`

File: src/Utilities/matrix_of_constraints.jl [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+    filter_constraints::Union{Nothing,Function} = nothing,
+)
+    set_number_of_columns(
+        dest.coefficients,
+        MOI.get(src, MOI.NumberOfVariables()),
+    )
+
+    for (F, S) in constraint_types
+        _allocate_constraints(dest, src, index_map, F, S, filter_constraints)
+    end
```
Comment (odow) [inline src/Utilities/matrix_of_constraints.jl:340]: `return`

File: src/Utilities/matrix_of_constraints.jl [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+    )
+    if filter_constraints !== nothing
+        filter!(filter_constraints, cis_src)
+    end
+    for ci_src in cis_src
+        func = MOI.get(src, MOI.CanonicalConstraintFunction(), ci_src)
+        set = MOI.get(src, MOI.ConstraintSet(), ci_src)
+        push!(model.are_indices_mapped, length(model.caches) + 1)
+        index_map[ci_src] = _add_constraint(model, i, index_map, func, set)
+    end
```
Comment (odow) [inline src/Utilities/matrix_of_constraints.jl:282]: `return`

File: src/Utilities/product_of_sets.jl [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+by [`set_types`](@ref)..
+"""
+set_index(sets::ProductOfSets, ::Type{S}) where {S<:MOI.AbstractSet} = nothing
+
+"""
+    set_types(sets::ProductOfSets)
+
+Return the list of the types of the sets allowed in the cartesian product.
+"""
+function set_types end
```
Comment (odow) [inline src/Utilities/product_of_sets.jl:21]: This really belong in `MatrixOfConstraints`.

File: src/Utilities/matrix_of_constraints.jl [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+        MOI.get(src, MOI.NumberOfVariables()),
+    )
+
+    for (F, S) in constraint_types
+        _allocate_constraints(dest, src, index_map, F, S, filter_constraints)
+    end
+end
+
+function final_touch(model::MatrixOfConstraints, index_map)
+    num_rows = number_of_rows(model.sets)
```
Comment (odow) [inline src/Utilities/matrix_of_constraints.jl:344]: This needs documenting

File: src/Utilities/matrix_of_constraints.jl [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+    resize!(model.constants, num_rows)
+    set_number_of_rows(model.coefficients, num_rows)
```
Comment (odow) [inline src/Utilities/matrix_of_constraints.jl:346]: This needs documenting

File: src/Utilities/matrix_of_constraints.jl [hunk 17]
```diff
@@ -0,0 +1,357 @@
+# Constants: stored in a `Vector` or `Box` or any other type implementing:
+# `empty!`, `resize!` and `load_constants`.
```
Comment (odow) [inline src/Utilities/matrix_of_constraints.jl:2]: This needs to be documented better.

File: src/Utilities/product_of_sets.jl [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+"""
+abstract type ProductOfSets{T} end
+
+"""
+    set_index(sets::ProductOfSets, ::Type{S}) where {S<:MOI.AbstractSet}
+
+Return an integer corresponding to the index of the set type in the list given
+by [`set_types`](@ref)..
+"""
+set_index(sets::ProductOfSets, ::Type{S}) where {S<:MOI.AbstractSet} = nothing
```
Comment (odow) [inline src/Utilities/product_of_sets.jl:14]: This needs to be documented in MatrixOfConstraints

File: src/Utilities/sparse_matrix.jl [hunk 19]
```diff
@@ -0,0 +1,118 @@
+import SparseArrays
+
+abstract type AbstractIndexing end
+struct ZeroBasedIndexing <: AbstractIndexing end
+struct OneBasedIndexing <: AbstractIndexing end
+
+first_index(::ZeroBasedIndexing) = 0
```
Comment (odow) [inline src/Utilities/sparse_matrix.jl:7]: This should all be private

**Review by odow — COMMENTED** [review]
This needs documentation added to the `Utilities` part of the manual.

I wonder about making a subdirectory in `src/Utilities` to contain these. That would isolate the code a little more, and better document what goes with what. 

I'm also keen to see what impact this has on our time-to-first-solve issue.

File: test/Utilities/copy.jl (hunk 6 above)
Comment (blegat) [reply test/Utilities/copy.jl:656]: This needs to be implemented by the type of the `constraints` field but it's currently not public API.
It's best we don't let user create new types at the moment until we converge. This is similar to `_delete_variables`, `_deleted_constraints` and `_throw_if_cannot_delete`.

**Review by odow — APPROVED** [review]
I have some reservations about adding yet more complexity to MOI. And it would be better if the documentation clearly identified which functions needed to be implemented for which bits, but it may be easier to merge and keep iterating in new PRs. We still have time to break things, so we should merge and then update solvers in branches to figure out what tweaks need to be made.

File: docs/src/submodules/Utilities/reference.md [hunk 20]
```diff
@@ -79,6 +79,47 @@ Utilities.load
...
+Utilities.set_number_of_rows
+Utilities.load_terms
+Utilities.final_touch
+```
+
+### Product of sets
+
+```@docs
+Utilities.ProductOfSets
+Utilities.set_index
```
[the remark this answers is not in the document]
Comment (blegat) [reply docs/src/submodules/Utilities/reference.md:114]: I made this into a well-defined API to make the code more understandable. We will need yet another one for list of sets that are not known at compile time, e.g., DiffOpt but we might have enough with 3, can't say for now.

File: src/Utilities/product_of_sets.jl [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+which allows a more efficient implementation.
+"""
+abstract type OrderedProductOfScalarSets{T} <: OrderedProductOfSets{T} end
+macro product_of_scalar_sets(name, set_types...)
+    esc_name = esc(name)
+    T = esc(:T)
+    type_def = :(struct $esc_name{$T} <: $MOIU.OrderedProductOfScalarSets{$T}
+        num_rows::Vector{Int}
+        function $esc_name{$T}() where {$T}
+            return new(zeros(Int, $(1 + length(set_types))))
```
[the remark this answers is not in the document]
Comment (blegat) [reply src/Utilities/product_of_sets.jl:171]: Yes, that not so much. I used to store the dimension for the scalar sets for `OrderedProductOfSets` so `OrderedProductOfScalarSets` was a big win but now even `OrderedProductOfSets` don't store the dimension of scalar sets so we could indeed get rid of `OrderedProductOfScalarSets` (although it's just a few lines of code).

**Comment by blegat** [comment]
> And it would be better if the documentation clearly identified which functions needed to be implemented for which bits, but it may be easier to merge and keep iterating in new PRs. We still have time to break things, so we should merge and then update solvers in branches to figure out what tweaks need to be made.

Yes I agree. It's ready so that people can start play with it and we can iterate a bit before releasing.

File: src/Utilities/product_of_sets.jl (hunk 21 above)
[the remark this answers is not in the document]
Comment (odow) [reply src/Utilities/product_of_sets.jl:171]: Part of this is that it's just a few lines of code, but much more code to compile.

[the remark this answers is not in the document]
Comment (blegat) [reply src/Utilities/product_of_sets.jl:171]: Yes, I think we can remove it
